text/0000-offset_of.md

```diff
@@ -0,0 +1,480 @@
+
+Use on other types is an error, although this may be relaxed in some cases in
+the future (see the [Future possibilities][future-possibilities] section).
+
+As a note: the implementation is strongly encouraged to not have runtime
+resource usage dependent on the values of `$Container` or `$field`. In
+particular, the implementation should not allocate space for an instance of
+`$Container` on the runtime stack.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. This exposes layout information at compile time which is otherwise not
+   exposed until runtime. This can cause compatibility hazards similar to
+   `mem::size_of` or `mem::align_of`, but plausibly greater as it provides even
+   more information.
+
+    That said, this API allows querying information which (if needed at compile
+    time) would otherwise be hard-coded, so in some cases it may reduce the risk
+    of a compatibility hazard.
+
+2. This is a low level feature that most code won't need to use, so perhaps it
+   is better off left out.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The general rationale is that it should remove the need to hardcode, hand-roll,
+or pull in a third-party crate in order to compute field offsets. This hopefully
+should remove as many barriers
+
+That said, there are several alternatives to this, some of which were even
+considered:
+
+1. Do nothing, and tell users to use the [`memoffset`][memoffset] crate, or to
+   hard-code constant offsets.
+
+    This was not chosen as this operation seems fundamental enough to provided
+    by the standard library, especially given how often it is incorrectly
+    implemented in the wild.
+
+2. Add `offset_of!`, but disallow use on `#[repr(Rust)]` types.
+
+    This would make `core::mem::offset_of!` have less functionality than the
+    implementation from `memoffset`, or the implementation they could implement
+    if they computed it manually.
+
+3. Require that all fields of `$Container` be visible at the invocation site,
+   rather than just requiring that `$field` is.
+
+    As above, this would make `core::mem::offset_of!` worse than the version
+    they'd have written themselves and/or an off-the-shelf implementation.
+
+4. Add `offset_of!`, but disallow use during constant evaluation.
+
+    This would mean that users which need const access to `offset_of!` must
+    continue to hardcode the field offsets as constants, which is undesirable,
+    error-prone, and can cause compatibility hazards.
+
+5. Try to make `addr_of!((*null::<$Container>()).$field) as usize` work for this:
+
+    Currently this is UB (due to dereferencing a null pointer) and does not
+    support use in const (due to accessing the address of a raw pointer).
+    Changing both of these issues would be challenging, but may be possible.
+
+    This was not chosen because seems difficult, and would be harder to teach
+    (or read) than `core::mem::offset_of`.
```

Comment:
FWIW my personal "ideal semantics" for Rust would allow this. But whether we can get those semantics [depends on LLVM](https://rust-lang.zulipchat.com/#narrow/stream/187780-t-compiler.2Fwg-llvm/topic/.60getelementptr.20nowrap.60)...

Reply:
IMO even if we allow this to work (it would be nice for `bindgen` output to be less full of UB...), it would still be worth providing `offset_of!`, as it's more readable, teachable, etc (As I mention).

That said, unless you *also* mean supporting `ptr.addr()` in const eval (which seems fraught with issues), this wouldn't really be a full replacement.

Reply:
Yeah, I agree offset_of should probably be provided by the language / standard library.

Reply:
I've elaborated on this a bit.